src/server/worker/pipeline/service/service.go

```diff
@@ -93,6 +96,30 @@ func forEachJob(pachClient *client.APIClient, pipelineInfo *pps.PipelineInfo, lo
 		if ji.State == pps.JobState_JOB_FINISHING {
 			return nil // don't pick up a "finishing" job
 		}
+		// Only restart the job once all source commits are closed.
+		// Otherwise a job will start for the service pipeline while without its complete set of data.
+		ci, err := pachClient.PfsAPIClient.InspectCommit(pachClient.Ctx(), &pfs.InspectCommitRequest{
+			Commit: &pfs.Commit{
+				Repo: &pfs.Repo{
+					Name: ji.Job.Pipeline.Name, Type: "user", Project: ji.Job.Pipeline.Project,
+				},
+				Id: ji.Job.Id,
+			},
+		})
+		if err != nil {
+			return err
+		}
+		log.Info(pachClient.Ctx(), "retrieved job commit")
+		for _, src := range ci.DirectProvenance {
+			log.Info(pachClient.Ctx(), "waiting on job's source commit to finish",
+				zap.String("job", ji.Job.String()),
+				zap.String("source_commit", src.String()))
+			if _, err := pachClient.PfsAPIClient.InspectCommit(pachClient.Ctx(), &pfs.InspectCommitRequest{
+				Commit: src, Wait: pfs.CommitState_FINISHED,
```

Comment:
this is a great point. It may not be possible today, but I think it's worth being protective here and added a mechanism to quit if we're in this step while another job comes in